Specs/GCM.lean

```diff
@@ -0,0 +1,198 @@
+
+protected theorem initialize_J0_simplification
+  (lv : Nat) (x : Nat) (h : lv ≤ x * 128):
+  lv + (x * 128 - lv + 64) + 64  = x * 128 + 128 := by omega
+
+protected def initialize_J0 (H : BitVec 128) (IV : BitVec lv) :=
+  if h₀ : lv == 96
+  then have h₁ : lv + 31 + 1 = 128 := by
+         simp only [Nat.succ.injEq]
+         exact Nat.eq_of_beq_eq_true h₀
+       h₁ ▸ (IV ++ BitVec.zero 31 ++ 0b1#1)
+  else let s := GCM.ceiling_in_bits lv - lv
+       have h₂ : 128 ∣ (lv + (s + 64) + 64) := by
+         simp only [s, GCM.ceiling_in_bits]
+         rw [GCM.initialize_J0_simplification lv (GCM.ceiling_in_blocks lv)
+             (by apply GCM.bits_le_ceiling_in_bits)]
+         omega
+       have h₃ : 8 ∣ (lv + (s + 64) + 64) := by
+         simp only [s, GCM.ceiling_in_bits]
+         rw [GCM.initialize_J0_simplification lv (GCM.ceiling_in_blocks lv)
+             (by apply GCM.bits_le_ceiling_in_bits)]
+         omega
+       let block := rev_elems (lv + (s + 64 ) + 64) 8
+                      (IV ++ (BitVec.zero (s + 64)) ++ (BitVec.ofNat 64 lv))
+                      h₃ (by decide)
+       rev_elems 128 8 (GHASH H block h₂) (by decide) (by decide)
+
+protected theorem GCM_AE_DE_simplification1 (a : Nat) (v : Nat) (p : Nat) (u : Nat) :
+  64 + 64 + u + p + v + a = 128 + (u + p) + (v + a) := by omega
+
+protected theorem GCM_AE_DE_simplification2
+  (y : Nat) (x : Nat) (h : y ≤ x * 128):
+  (x * 128 - y) + y = x * 128 := by omega
+
+/-- GCM_AE : Galois Counter Mode Authenticated Encryption -/
+def GCM_AE {m : Nat} (CIPH : Cipher (n := 128) (m := m))
+  (K : BitVec m) (IV : BitVec lv) (P : BitVec p) (A : BitVec a) (t : Nat)
+  : (BitVec p) × (BitVec t) :=
+  let H := CIPH (BitVec.zero 128) K
+  let J0 : BitVec 128 := GCM.initialize_J0 H IV
+  let ICB := inc_s 32 J0 (by decide) (by decide)
+  let C := GCTR (m := m) CIPH K ICB P
+  let u := GCM.ceiling_in_bits p - p
+  let v := GCM.ceiling_in_bits a - a
+  let block := rev_elems 64 8 (BitVec.ofNat 64 p) (by decide) (by decide)
+               ++ rev_elems 64 8 (BitVec.ofNat 64 a) (by decide) (by decide)
+               ++ BitVec.zero u ++ C ++ BitVec.zero v ++ A
+  have h : 128 ∣ 64 + 64 + u + p + v + a := by
+    rw [GCM.GCM_AE_DE_simplification1]
+    simp only [u, v]
+    simp only [GCM.ceiling_in_bits]
+    rw [GCM.GCM_AE_DE_simplification2 p (GCM.ceiling_in_blocks p)
+        (by apply GCM.bits_le_ceiling_in_bits)]
+    rw [GCM.GCM_AE_DE_simplification2 a (GCM.ceiling_in_blocks a)
+        (by apply GCM.bits_le_ceiling_in_bits)]
+    omega
+  let S := GHASH H block h
+  let T := truncate t $ GCTR (m := m) CIPH K J0 S
+  (C, T)
+
+def length_constraints (_IV : BitVec v) (_A : BitVec a) (_C : BitVec c)
+  : Bool :=
+     1 ≤ v && v ≤ 2^64 - 1
+  && c ≤ 2^39 - 256
+  && a ≤ 2^64 - 1
+
+/-- GCM_AD : Galois Counter Mode Authenticated Decryption
+    GCM_AD returns none when decryption fails. -/
+def GCM_AD {m : Nat} (CIPH : Cipher (n := 128) (m := m))
```

Comment:
I removed parameter `m` instead of adding `t`. The function signature also contains other parameters like `lv`, `a` and etc. It would be tedious to add all.